File: services/payment-gateway-go/internal/observability/logger.go
package observability
import ("encoding/json"; "log"; "strings")
type Logger struct{service string; env string; version string}
func NewLogger(service, env, version string) *Logger {return &Logger{service:service,env:env,version:version}}
func (l *Logger) Info(msg string, fields map[string]interface{}){l.log("info",msg,fields)}
func (l *Logger) Error(msg string, fields map[string]interface{}){l.log("error",msg,fields)}
func (l *Logger) Audit(action string, fields map[string]interface{}){fields["action"]=action; l.log("audit",action,fields)}
func (l *Logger) log(level, msg string, fields map[string]interface{}){
    if fields==nil{fields=make(map[string]interface{})}
    fields["level"]=level; fields["service"]=l.service; fields["env"]=l.env; fields["version"]=l.version; fields["message"]=l.redact(msg)
    for k,v:=range fields{if isSensitive(k){fields[k]="***REDACTED***"} else if s,ok:=v.(string); ok&&isSensitiveValue(s){fields[k]="***REDACTED***"}}
    b,_:=json.Marshal(fields); log.Println(string(b))
}
func (l *Logger) redact(s string) string{
    sensitive:=[]string{"password","secret","token","jwt","DATABASE_URL","REDIS_URL"}
    for _,key:=range sensitive{if strings.Contains(strings.ToLower(s),strings.ToLower(key)){return "***REDACTED***"}}
    return s
}
func (l *Logger) WithRequestID(requestID string) *Logger {return l}
func isSensitive(key string) bool{
    lower:=strings.ToLower(key)
    sensitive:=[]string{"password","secret","token","jwt","api_key","private_key","database_url","redis_url"}
    for _,s:=range sensitive{if strings.Contains(lower,s){return true}}
    return false
}
func isSensitiveValue(v string) bool{if len(v)>20&&(strings.HasPrefix(v,"eyJ")||strings.HasPrefix(v,"sk_")){return true}; return false}

File: services/security-rust/src/observability/logger.rs
use serde_json::Value;

#[derive(Clone)]
pub struct Logger {
    service: String,
    env: String,
    version: String,
}

impl Logger {
    pub fn new(service: &str, env: &str, version: &str) -> Self {
        Self{ service: service.to_string(), env: env.to_string(), version: version.to_string() }
    }
    pub fn info(&self, msg: &str, fields: Value) {
        let log = serde_json::json!({
            "level": "info",
            "service": self.service,
            "env": self.env,
            "version": self.version,
            "message": self.redact(msg),
            "fields": self.redact_value(fields)
        });
        println!("{}", log);
    }
    pub fn error(&self, msg: &str, fields: Value) {
        let log = serde_json::json!({
            "level": "error",
            "service": self.service,
            "env": self.env,
            "version": self.version,
            "message": self.redact(msg),
            "fields": self.redact_value(fields)
        });
        eprintln!("{}", log);
    }
    pub fn audit(&self, action: &str, fields: Value) {
        let log = serde_json::json!({
            "level": "audit",
            "service": self.service,
            "action": action,
            "fields": self.redact_value(fields)
        });
        println!("{}", log);
    }
    fn redact(&self, s: &str) -> String {
        let lower = s.to_lowercase();
        if lower.contains("password") || lower.contains("secret") || lower.contains("token") || lower.contains("jwt") {
            "***REDACTED***".to_string()
        } else {
            s.to_string()
        }
    }
    fn redact_value(&self, value: Value) -> Value {
        match value {
            Value::Object(map) => {
                let mut new_map = serde_json::Map::new();
                for (k, v) in map {
                    let lower = k.to_lowercase();
                    if lower.contains("password") || lower.contains("secret") || lower.contains("token") || lower.contains("jwt") || lower.contains("api_key") {
                        new_map.insert(k, Value::String("***REDACTED***".to_string()));
                    } else {
                        new_map.insert(k, self.redact_value(v));
                    }
                }
                Value::Object(new_map)
            }
            _ => value,
        }
    }
}
